heartbeat/processor/peerAuthenticationRequestsProcessor.go

```diff
@@ -0,0 +1,234 @@
+func NewPeerAuthenticationRequestsProcessor(args ArgPeerAuthenticationRequestsProcessor) (*PeerAuthenticationRequestsProcessor, error) {
+	err := checkArgs(args)
+	if err != nil {
+		return nil, err
+	}
+
+	processor := &PeerAuthenticationRequestsProcessor{
+		requestHandler:         args.RequestHandler,
+		nodesCoordinator:       args.NodesCoordinator,
+		peerAuthenticationPool: args.PeerAuthenticationPool,
+		shardId:                args.ShardId,
+		epoch:                  args.Epoch,
+		messagesInChunk:        args.MessagesInChunk,
+		minPeersThreshold:      args.MinPeersThreshold,
+		delayBetweenRequests:   args.DelayBetweenRequests,
+		maxTimeout:             args.MaxTimeout,
+	}
+
+	var ctx context.Context
+	ctx, processor.cancel = context.WithCancel(context.Background())
+
+	go processor.startRequestingMessages(ctx)
+
+	return processor, nil
+}
+
+func checkArgs(args ArgPeerAuthenticationRequestsProcessor) error {
+	if check.IfNil(args.RequestHandler) {
+		return heartbeat.ErrNilRequestHandler
+	}
+	if check.IfNil(args.NodesCoordinator) {
+		return heartbeat.ErrNilNodesCoordinator
+	}
+	if check.IfNil(args.PeerAuthenticationPool) {
+		return heartbeat.ErrNilPeerAuthenticationPool
+	}
+	if args.MessagesInChunk < minMessagesInChunk {
+		return fmt.Errorf("%w for MessagesInChunk, provided %d, min expected %d",
+			heartbeat.ErrInvalidValue, args.MessagesInChunk, minMessagesInChunk)
+	}
+	if args.MinPeersThreshold < minMessagesThreshold {
+		return fmt.Errorf("%w for MinPeersThreshold, provided %f, min expected %f",
+			heartbeat.ErrInvalidValue, args.MinPeersThreshold, minMessagesThreshold)
+	}
+	if args.DelayBetweenRequests < minDelayBetweenRequests {
+		return fmt.Errorf("%w for DelayBetweenRequests, provided %d, min expected %d",
+			heartbeat.ErrInvalidTimeDuration, args.DelayBetweenRequests, minDelayBetweenRequests)
+	}
+	if args.MaxTimeout < minTimeout {
+		return fmt.Errorf("%w for MaxTimeout, provided %d, min expected %d",
+			heartbeat.ErrInvalidTimeDuration, args.MaxTimeout, minTimeout)
+	}
+
+	return nil
+}
+
+func (processor *PeerAuthenticationRequestsProcessor) startRequestingMessages(ctx context.Context) {
+	defer func() {
+		if processor.cancel != nil {
+			processor.cancel()
+		}
+	}()
+
+	sortedValidatorsKeys, err := processor.getSortedValidatorsKeys()
+	if err != nil {
+		return
+	}
+
+	// first request messages by chunks
+	processor.requestKeysChunks(sortedValidatorsKeys)
+
+	// start endless loop until enough messages received or timeout reached
+	requestsTimer := time.NewTimer(processor.delayBetweenRequests)
+	timeoutTimer := time.NewTimer(processor.maxTimeout)
+	for {
+		if processor.isThresholdReached(sortedValidatorsKeys) {
+			log.Debug("received enough messages, closing PeerAuthenticationRequestsProcessor go routine")
+			return
+		}
+
+		requestsTimer.Reset(processor.delayBetweenRequests)
+		select {
+		case <-requestsTimer.C:
+			processor.requestMissingKeys(sortedValidatorsKeys)
+		case <-timeoutTimer.C:
```

Comment:
this can be safely removed and instead use another type of context on the constructor:
```GO
ctx, processor.cancel = context.WithTimeout(context.Background(), args.MaxTimeout)
```